pkg/projectrequestlimit/validation_test.go

```diff
@@ -0,0 +1,414 @@
+	"k8s.io/client-go/tools/cache"
+	"sigs.k8s.io/yaml"
+)
+
+func TestMaxProjectByRequester(t *testing.T) {
+	config := multiLevelConfig()
+	tests := []struct {
+		userLabels      map[string]string
+		expectUnlimited bool
+		expectedLimit   int64
+	}{
+		{
+			userLabels:      map[string]string{"platinum": "yes"},
+			expectUnlimited: true,
+		},
+		{
+			userLabels:    map[string]string{"gold": "yes"},
+			expectedLimit: 10,
+		},
+		{
+			userLabels:    map[string]string{"silver": "yes", "bronze": "yes"},
+			expectedLimit: 3,
+		},
+		{
+			userLabels:    map[string]string{"unknown": "yes"},
+			expectedLimit: 1,
+		},
+	}
+
+	for _, tc := range tests {
+		userLister := fakeUserLister([]*userapi.User{
+			fakeUser("testuser", tc.userLabels),
+		})
+		reqLimit := projectRequestLimitValidator{
+			userLister: userLister,
+		}
+		maxProjects, hasLimit, err := reqLimit.maxProjectsByRequester("testuser", config)
+		if err != nil {
+			t.Errorf("Unexpected error: %v", err)
+		}
+		if tc.expectUnlimited {
+			if hasLimit {
+				t.Errorf("Expected no limit, but got limit for labels %v", tc.userLabels)
+			}
+			continue
+		}
+		if !tc.expectUnlimited && !hasLimit {
+			t.Errorf("Did not expect unlimited for labels %v", tc.userLabels)
+			continue
+		}
+		if maxProjects != tc.expectedLimit {
+			t.Errorf("Did not get expected limit for labels %v. Got: %d. Expected: %d", tc.userLabels, maxProjects, tc.expectedLimit)
+		}
+	}
+}
+
+func TestProjectCountByRequester(t *testing.T) {
+	nsLister := fakeNamespaceLister(map[string]projectCount{
+		"user1": {1, 5}, // total 6, expect 4
+		"user2": {5, 1}, // total 6, expect 5
+		"user3": {1, 0}, // total 1, expect 1
+	})
+	reqLimit := &projectRequestLimitValidator{
+		nsLister: nsLister,
+	}
+	tests := []struct {
+		user   string
+		expect int
+	}{
+		{
+			user:   "user1",
+			expect: 4,
+		},
+		{
+			user:   "user2",
+			expect: 5,
+		},
+		{
+			user:   "user3",
+			expect: 1,
+		},
+	}
+
+	for _, test := range tests {
+		actual, err := reqLimit.projectCountByRequester(test.user)
+		if err != nil {
+			t.Errorf("unexpected: %v", err)
+		}
+		if actual != int64(test.expect) {
+			t.Errorf("user %s got %d, expected %d", test.user, actual, test.expect)
+		}
+	}
+
+}
+
+func TestAdmit(t *testing.T) {
+	nsLister := fakeNamespaceLister(map[string]projectCount{
+		"user1": {0, 1},
+		"user2": {2, 2},
+		"user3": {5, 3},
+		"user4": {1, 0},
+
+		"system:admin": {5, 0},
+	})
+
+	userLister := fakeUserLister([]*userapi.User{
+		fakeUser("user2", map[string]string{"bronze": "yes"}),
+		fakeUser("user3", map[string]string{"platinum": "yes"}),
+		fakeUser("user4", map[string]string{"unknown": "yes"}),
+	})
+
+	reqLimit := projectRequestLimitValidator{
+		userLister: userLister,
+		nsLister:   nsLister,
+	}
+
+	tests := []struct {
+		config          *projectv1.ProjectRequestLimit
+		user            string
+		expectForbidden bool
+	}{
+		// {
+		// 	config: multiLevelConfig(),
+		// 	user:   "user1",
+		// },
+		{
```

Comment:
I believe that's because the old tests used this:
https://github.com/openshift/openshift-apiserver/blob/6ae368b1f0170d8b96dcc926fe13a541ba6e3bd1/pkg/project/apiserver/admission/requestlimit/admission_test.go#L276
https://github.com/openshift/openshift-apiserver/blob/6ae368b1f0170d8b96dcc926fe13a541ba6e3bd1/pkg/project/apiserver/admission/requestlimit/admission_test.go#L388
Here the default user is returned without labels, where as in this test you create the users but not all of them.

Reply:
I didn't know the fake client will always respond with some `User{}`/resource, thank you. 

In that case I'd say the original test is wrong and so is the assumption that `ProjectRequests` for non-existing users should be allowed.

Reply:
just one comment here from my side, we're building this for people upgrading from 3.11 to 4.x. So even if there are bugs we should keep the behaviour consistent with 3.11 to not cause any migration issues for our customers.

Reply:
Just to clarify the bug(actually it's not a bug, just non-obvious behavior) is in the test.